lazy-reducer-enhancer.d.ts

```diff
@@ -0,0 +1,69 @@
+import {
+  Reducer,
+  Store,
+  ReducersMapObject,
+  Action,
+  AnyAction,
+  StoreEnhancer,
+  StoreEnhancerStoreCreator,
+  combineReducers
+} from "redux";
+
+export interface ICombineReducers<S> {
+  (reducers: ReducersMapObject<S, any>): Reducer<S>;
+}
+
+export interface ICombineReducers<S, A extends Action = AnyAction> {
+  (reducers: ReducersMapObject<S, A>): Reducer<S, A>;
+}
+
+/**
+  If you are lazy loading any connected elements, then these elements must be
+  able to lazily install their reducers. This is a store enhancer that
+  enables that.
+
+  @template S The type of state held by this store.
+  @template A the type of actions which may be dispatched by this store.
+  @template Ext Store extension that is mixed into the Store type.
+  @template StateExt State extension that is mixed into the state type.
+
+  @return enhanced store with a `addReducers` method.
+  
+  @example
+  // Sample use (where you define your redux store, in store.js):
+  import lazyReducerEnhancer from '../node_modules/pwa-helpers/lazy-reducer-enhancer.js';
+  import someReducer from './reducers/someReducer.js';
+
+  export const store = createStore(
+    (state, action) => state,
+    compose(lazyReducerEnhancer(combineReducers), applyMiddleware(thunk))
+  );
+
+  Then, in your page/element, you can lazy load a specific reducer with:
+  store.addReducers({
+    someReducer
+  });
+*/
+export declare function lazyReducerEnhancer<
+  S,
+  A extends Action = AnyAction,
+  Ext = {},
+  StateExt = {}
```

Comment:
I had trouble getting the typings to work correctly when using _both_ redux-thunk _and_ lazy-reducer enhancers (i.e. `compose(lazyReducerEnhancer(combineReducers), applyMiddleware(thunk))`). I [ended up removing lazy-reducer-enhancer entirely](https://github.com/Polymer/pwa-starter-kit/commit/869f01a990ec357ee1d65ec82125cff5abfc3588) and just exported an `addReducers` function from store.js which calls `store.replaceReducer()` itself. Looking at [your fork](https://github.com/e2forks/pwa-starter-kit/tree/template-typescript), seems like you had to do a similar workaround, right?

Reply:
yup. I have the same problem. 

My solution is kinda similar. The only difference is that i wrapped the `addReducers` function to return a `store` that is typed cast to the relevant `State` and `Action` types based on the `reducerMapObject` argument - so that I do not need to have a "global" state.

The issue is with the type definition for `compose` - it doesn't merge `Ext` and `StateExt`, and the last func has no awareness of the prior functions. I could possibly write another set of `compose` declaration where the arguments are `StoreEnhancers` instead of generic functions. This will probably solve that.

```
export function compose(
  f1: StoreEnhancer<Ext0, StateExt0>, f2: StoreEnhancer<Ext1, StateExt1>
): StoreEnhancer<Ext0 & Ext1, StateExt0 & StateExt1>;
```
But it is not really scalable as there are so many combinations.

Reply:
This could actually be fine for pwa-starter-kit - we could require users to specify this typing in their store.js based on however many args they use: https://github.com/Polymer/pwa-starter-kit/commit/7a4b58f15d0df315543bfaae1bc80263339b1c2c#diff-cfb8e44b6757fd740446e9e4486b73c1R47